# EDA: Diagnosing Diabetes

In this project, you'll imagine you are a data scientist interested in exploring data that looks at how certain diagnostic factors affect the diabetes outcome of women patients.

You will use your EDA skills to help inspect, clean, and validate the data.

**Note**: This [dataset](https://www.kaggle.com/uciml/pima-indians-diabetes-database) is from the National Institute of Diabetes and Digestive and Kidney Diseases. It contains the following columns:

- `Pregnancies`: Number of times pregnant
- `Glucose`: Plasma glucose concentration per 2 hours in an oral glucose tolerance test
- `BloodPressure`: Diastolic blood pressure
- `SkinThickness`: Triceps skinfold thickness
- `Insulin`: 2-Hour serum insulin
- `BMI`: Body mass index
- `DiabetesPedigreeFunction`: Diabetes pedigree function
- `Age`: Age (years)
- `Outcome`: Class variable (0 or 1)

Let's get started!

## Initial Inspection

1. First, familiarize yourself with the dataset [here](https://www.kaggle.com/uciml/pima-indians-diabetes-database).

   Look at each of the nine columns in the documentation.
   
   What do you expect each data type to be?

Expected data type for each column:

- `Pregnancies`: int64 (count)
- `Glucose`: int64 (measurement, unit unknown)
- `BloodPressure`: int64 (measurement in mm Hg)
- `SkinThickness`: int64 (measurement in mm)
- `Insulin`: int64 (measurement in mu U/ml)
- `BMI`: float64 (calculated as weights in kg/(height in m)^2
- `DiabetesPedigreeFunction`: float64 (calculated?)
- `Age`: int64 (years)
- `Outcome`: int64 (0 or 1), can be converted to boolean (True or False)

2. Next, let's load in the diabetes data to start exploring.

   Load the data in a variable called `diabetes_data` and print the first few rows.
   
   **Note**: The data is stored in a file called `diabetes.csv`.

In [1]:
import pandas as pd
import numpy as np

# load in data
diabetes_data = pd.read_csv('diabetes.csv')
diabetes_data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age Outcome  
0                     0.627   50       1  
1                     0.351   31       0  
2                     0.672   32       1  
3                     0.167   21       0  
4                     2.288   33       1

3. How many columns (features) does the data contain?

In [2]:
# print number of columns
diabetes_data.shape[1] # or:
#print(len(diabetes_data.columns))

9

4. How many rows (observations) does the data contain?

In [3]:
# print number of rows
diabetes_data.shape[0] # or:
#print(len(diabetes_data))

768

## Further Inspection

5. Let's inspect `diabetes_data` further.

   Do any of the columns in the data contain null (missing) values?

In [4]:
# find whether columns contain null values
diabetes_data.info() # or:
print(diabetes_data.isnull().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    object 
dtypes: float64(2), int64(6), object(1)
memory usage: 54.1+ KB
Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age               

Not columns with null values at the first glance.

6. If you answered no to the question above, not so fast!

   While it's technically true that none of the columns contain null values, that doesn't necessarily mean that the data isn't missing any values.
   
   When exploring data, you should always question your assumptions and try to dig deeper.
   
   To investigate further, calculate summary statistics on `diabetes_data` using the `.describe()` method.

In [5]:
# perform summary statistics
diabetes_data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age  
count  768.000000                768.000000  768.000000  
mean    31.992578                  0.471876   33.240885  
std      7.884160                  0.331329   11.760232  
min      0.000000                  0.078000   21.000000  
25%     27.300000                  0.243750   24.000000  
50%     32.000000                  0.372500   29.000000  
75%     36.600000                  0.626250   41.000000  
max     67.100000                  2.420000   81.000000

7. Looking at the summary statistics, do you notice anything odd about the following columns?

   - `Glucose`
   - `BloodPressure`
   - `SkinThickness`
   - `Insulin`
   - `BMI`

**Your response to question 7**:
yes, the minimal values are 0!

8. Do you spot any other outliers in the data?

**Your response to question 8**:
Some variables have too high maximum values, much higher than the mean or 75%-ile:
* Pregnancies: 17 >> 6 > 3
* Glucose: 199 > 140.25 > 117 (can it be so high?)
* BloodPressure: 122 > 80 > 72 (sistolic instead of diastilic measured?)
* SkinThickness: 99 >> 32 > 23
* Insulin: 846 >> 127.25 > 30.5
* BMI: 67.1 >> 36.6 > 32

9. Let's see if we can get a more accurate view of the missing values in the data.

   Use the following code to replace the instances of `0` with `NaN` in the five columns mentioned:
   
   ```py
   diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']] = diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']].replace(0, np.NaN)
   ```

In [6]:
# replace instances of 0 with NaN
diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']] = \
diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']].replace(0, np.NaN)

10. Next, check for missing (null) values in all of the columns just like you did in Step 5.

    Now how many missing values are there?

In [7]:
# find whether columns contain null values after replacements are made
diabetes_data.info()

print('Missing values:')
print('Glucose: ', (768 - 763))
print('BloodPressure: ', (768 - 733))
print('SkinThickness: ', (768 - 541))
print('Insulin: ', (768 - 394))
print('BMI: ', (768 - 757))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   763 non-null    float64
 2   BloodPressure             733 non-null    float64
 3   SkinThickness             541 non-null    float64
 4   Insulin                   394 non-null    float64
 5   BMI                       757 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    object 
dtypes: float64(6), int64(2), object(1)
memory usage: 54.1+ KB
Missing values:
Glucose:  5
BloodPressure:  35
SkinThickness:  227
Insulin:  374
BMI:  11


In [8]:
# shorter way:
print(diabetes_data.isnull().sum())

Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64


11. Let's take a closer look at these rows to get a better idea of _why_ some data might be missing.

    Print out all the rows that contain missing (null) values.

In [9]:
# print rows with missing values
# Set the display option to show all 376 rows
pd.set_option('display.max_rows', 376)
diabetes_data[diabetes_data.isnull().any(axis=1)]

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6    148.0           72.0           35.0      NaN  33.6   
1              1     85.0           66.0           29.0      NaN  26.6   
2              8    183.0           64.0            NaN      NaN  23.3   
5              5    116.0           74.0            NaN      NaN  25.6   
7             10    115.0            NaN            NaN      NaN  35.3   
9              8    125.0           96.0            NaN      NaN   NaN   
10             4    110.0           92.0            NaN      NaN  37.6   
11            10    168.0           74.0            NaN      NaN  38.0   
12            10    139.0           80.0            NaN      NaN  27.1   
15             7    100.0            NaN            NaN      NaN  30.0   
17             7    107.0           74.0            NaN      NaN  29.6   
21             8     99.0           84.0            NaN      NaN  35.4   
22             7    196.0           90.0            NaN      NaN  39.8   
23             9    119.0           80.0           35.0      NaN  29.0   
26             7    147.0           76.0            NaN      NaN  39.4   
29             5    117.0           92.0            NaN      NaN  34.1   
30             5    109.0           75.0           26.0      NaN  36.0   
33             6     92.0           92.0            NaN      NaN  19.9   
34            10    122.0           78.0           31.0      NaN  27.6   
36            11    138.0           76.0            NaN      NaN  33.2   
37             9    102.0           76.0           37.0      NaN  32.9   
38             2     90.0           68.0           42.0      NaN  38.2   
41             7    133.0           84.0            NaN      NaN  40.2   
42             7    106.0           92.0           18.0      NaN  22.7   
44             7    159.0           64.0            NaN      NaN  27.4   
45             0    180.0           66.0           39.0      NaN  42.0   
46             1    146.0           56.0            NaN      NaN  29.7   
47             2     71.0           70.0           27.0      NaN  28.0   
48             7    103.0           66.0           32.0      NaN  39.1   
49             7    105.0            NaN            NaN      NaN   NaN   
55             1     73.0           50.0           10.0      NaN  23.0   
58             0    146.0           82.0            NaN      NaN  40.5   
60             2     84.0            NaN            NaN      NaN   NaN   
61             8    133.0           72.0            NaN      NaN  32.9   
62             5     44.0           62.0            NaN      NaN  25.0   
64             7    114.0           66.0            NaN      NaN  32.8   
65             5     99.0           74.0           27.0      NaN  29.0   
66             0    109.0           88.0           30.0      NaN  32.5   
67             2    109.0           92.0            NaN      NaN  42.7   
72            13    126.0           90.0            NaN      NaN  43.4   
74             1     79.0           75.0           30.0      NaN  32.0   
75             1      NaN           48.0           20.0      NaN  24.7   
76             7     62.0           78.0            NaN      NaN  32.6   
77             5     95.0           72.0           33.0      NaN  37.7   
78             0    131.0            NaN            NaN      NaN  43.2   
79             2    112.0           66.0           22.0      NaN  25.0   
80             3    113.0           44.0           13.0      NaN  22.4   
81             2     74.0            NaN            NaN      NaN   NaN   
83             0    101.0           65.0           28.0      NaN  24.6   
84             5    137.0          108.0            NaN      NaN  48.8   
86            13    106.0           72.0           54.0      NaN  36.6   
89             1    107.0           68.0           19.0      NaN  26.5   
90             1     80.0           55.0            NaN      NaN  19.1   
93       

12. Go through the rows with missing data. Do you notice any patterns or overlaps between the missing data?

**Your response to question 12**:
Almost all rows where 'Insulin' is missing, also has at least one missing value in another column.


13. Next, take a closer look at the data types of each column in `diabetes_data`.

    Does the result match what you would expect?

In [10]:
# print data types using .info() method
diabetes_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   763 non-null    float64
 2   BloodPressure             733 non-null    float64
 3   SkinThickness             541 non-null    float64
 4   Insulin                   394 non-null    float64
 5   BMI                       757 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    object 
dtypes: float64(6), int64(2), object(1)
memory usage: 54.1+ KB


14. To figure out why the `Outcome` column is of type `object` (string) instead of type `int64`, print out the unique values in the `Outcome` column.

In [11]:
# print unique values of Outcome column
print(diabetes_data.Outcome.unique())

['1' '0' 'O']


15. How might you resolve this issue?

**Your response to question 15**:
Some values in 'Outcome' column are characters 'O' in additon to '1' and '0'. Therefore, the type is 'object' instead of 'int'. According to the documentation, the value of this column should either be a 0 or a 1, so it is likely that instances of the character 'O' are typos.

I would check how many 'O' there are, then replace them with 0 and finally convert the column type to int.

## Next Steps:

16. Congratulations! In this project, you saw how EDA can help with the initial data inspection and cleaning process. This is an important step as it helps to keep your datasets clean and reliable.

    Here are some ways you might extend this project if you'd like:
    - Use `.value_counts()` to more fully explore the values in each column.
    - Investigate other outliers in the data that may be easily overlooked.
    - Instead of changing the `0` values in the five columns to `NaN`, try replacing the values with the median or mean of each column.

In [12]:
print(diabetes_data.Outcome.value_counts())

Outcome
0    494
1    268
O      6
Name: count, dtype: int64


In [13]:
diabetes_data.Outcome = diabetes_data.Outcome.replace('O', 0)
print(diabetes_data.Outcome.value_counts())

Outcome
0    494
1    268
0      6
Name: count, dtype: int64


In [14]:
diabetes_data.Outcome = diabetes_data.Outcome.astype(int)
print(diabetes_data.dtypes)

Pregnancies                   int64
Glucose                     float64
BloodPressure               float64
SkinThickness               float64
Insulin                     float64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object


In [15]:
diabetes_data[diabetes_data['Pregnancies'] == 17] # may be true, the women is 47 years old

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
159           17    163.0           72.0           41.0    114.0  40.9   

     DiabetesPedigreeFunction  Age  Outcome  
159                     0.817   47        1

In [16]:
diabetes_data[diabetes_data['Glucose'] == 199] # maybe a valid measurement, since she has diabetes

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
661            1    199.0           76.0           43.0      NaN  42.9   

     DiabetesPedigreeFunction  Age  Outcome  
661                     1.394   22        1

In [17]:
diabetes_data[diabetes_data['BloodPressure'] == 122] # maybe measured after she was running?

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
106            1     96.0          122.0            NaN      NaN  22.4   

     DiabetesPedigreeFunction  Age  Outcome  
106                     0.207   27        0

In [18]:
diabetes_data[diabetes_data['SkinThickness'] == 99] # prabably on outlier, since women with a higher BMI have lower skin thickness

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
579            2    197.0           70.0           99.0      NaN  34.7   

     DiabetesPedigreeFunction  Age  Outcome  
579                     0.575   62        1

In [19]:
diabetes_data[diabetes_data['Insulin'] == 846] # outlier based on mean and 75%-ile insulin values

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
13            1    189.0           60.0           23.0    846.0  30.1   

    DiabetesPedigreeFunction  Age  Outcome  
13                     0.398   59        1

In [20]:
diabetes_data[diabetes_data['BMI'] == 67.1] # probably a miscalculation

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
177            0    129.0          110.0           46.0    130.0  67.1   

     DiabetesPedigreeFunction  Age  Outcome  
177                     0.319   26        1

In [21]:
# assuming the patient' height is 1.62 m, she would weight 176 kg!
67.1 * 1.62**2

176.09724000000003

In [22]:
# replace instances of NaN (previously 0) with the mean of each column
for column in ['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']:
    diabetes_data[column] = diabetes_data[column].replace(0, diabetes_data[column].mean())
    
diabetes_data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  763.000000     733.000000     541.000000  394.000000   
mean      3.845052  121.686763      72.405184      29.153420  155.548223   
std       3.369578   30.535641      12.382158      10.476982  118.775855   
min       0.000000   44.000000      24.000000       7.000000   14.000000   
25%       1.000000   99.000000      64.000000      22.000000   76.250000   
50%       3.000000  117.000000      72.000000      29.000000  125.000000   
75%       6.000000  141.000000      80.000000      36.000000  190.000000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  757.000000                768.000000  768.000000  768.000000  
mean    32.457464                  0.471876   33.240885    0.348958  
std      6.924988                  0.331329   11.760232    0.476951  
min     18.200000                  0.078000   21.000000    0.000000  
25%     27.500000                  0.243750   24.000000    0.000000  
50%     32.300000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000